File: apps/web/src/utils/binderValueSnapshot.js
/**
 * Quantity-weighted Binder totals for the value overlay.
 * Golden cases: packages/contracts/binder_value_snapshot.json
 */

export const isPricedField = (value) => {
    if (value === null || value === undefined || value === '') return false;
    const n = Number(value);
    return n !== 0 && !Number.isNaN(n);
};

const firstPriced = (values) => {
    for (const value of values) {
        if (isPricedField(value)) return Number(value);
    }
    return null;
};

const foilAware = (row, foilKey, baseKey) =>
    row.isFoil ? firstPriced([row[foilKey], row[baseKey]]) : firstPriced([row[baseKey]]);

const sourceValue = (row, source, headline) => {
    if (headline === 'tcgMarketOnly') return firstPriced([row.tcgMarket]);
    if (source === 'cardmarket') {
        if (row.isFoil) {
            return firstPriced([row.cmTrendFoil, row.cmTrend, row.cmAvg, row.cmLow]);
        }
        return firstPriced([row.cmTrend, row.cmAvg, row.cmLow]);
    }
    return firstPriced([row.tcgMarket, row.tcgLow, row.tcgMid, row.tcgHigh]);
};

const sumField = (rows, pick) => {
    let amount = 0;
    let pricedCopies = 0;
    let unpricedCopies = 0;
    let anyPriced = false;
    for (const row of rows) {
        const value = pick(row);
        if (isPricedField(value)) {
            amount += Number(value) * row.quantity;
            pricedCopies += row.quantity;
            anyPriced = true;
        } else {
            unpricedCopies += row.quantity;
        }
    }
    return {
        amount: anyPriced ? amount : null,
        pricedCopies,
        unpricedCopies,
    };
};

/**
 * Map a web Binder row + live catalog Printing into the contract entry shape.
 * Missing/zero catalog numbers stay nullish so the snapshot can treat them as unpriced.
 */
export function snapshotEntryFromBinder(entry, live) {
    const card = live || entry.card || {};
    const finish =
        card.subTypeName ||
        card.finish ||
        (card.isFoil ? 'Foil' : 'Normal');
    const isFoil = Boolean(card.isFoil) || /foil/i.test(String(finish));
    const numOrNull = (...keys) => {
        for (const key of keys) {
            if (card[key] === null || card[key] === undefined || card[key] === '') continue;
            const n = Number(card[key]);
            if (!Number.isNaN(n)) return n;
        }
        return null;
    };
    return {
        printingId: entry.cardId || card._uniqueId || card.id,
        name: card.name || 'Unknown card',
        finish,
        isFoil,
        quantity: entry.quantity || 1,
        tcgMarket: numOrNull('marketPrice', 'tcgMarket', 'tcg_market'),
        tcgLow: numOrNull('lowPrice', 'tcgLow', 'tcg_low'),
        tcgMid: numOrNull('midPrice', 'tcgMid', 'tcg_mid'),
        tcgHigh: numOrNull('highPrice', 'tcgHigh', 'tcg_high'),
        cmTrend: numOrNull('cardmarketTrend', 'cmTrend', 'cm_trend'),
        cmLow: numOrNull('cardmarketLow', 'cmLow', 'cm_low'),
        cmAvg: numOrNull('cardmarketAvg', 'cmAvg', 'cm_avg'),
        cmTrendFoil: numOrNull('cardmarketTrendFoil', 'cmTrendFoil', 'cm_trend_foil'),
        cmLowFoil: numOrNull('cardmarketLowFoil', 'cmLowFoil', 'cm_low_foil'),
        cmAvgFoil: numOrNull('cardmarketAvgFoil', 'cmAvgFoil', 'cm_avg_foil'),
    };
}

/**
 * @param {Array<object>} entries contract-shaped rows (or already mapped)
 * @param {{ source?: 'tcgplayer'|'cardmarket', headline?: 'pricingValue'|'tcgMarketOnly' }} [options]
 */
export function buildBinderValueSnapshot(
    entries,
    { source = 'tcgplayer', headline = 'tcgMarketOnly' } = {},
) {
    const rows = (entries || []).filter((row) => (row.quantity || 0) >= 1);

    let copies = 0;
    let foilCopies = 0;
    let tcgUnpricedCopies = 0;
    let cmUnpricedCopies = 0;
    for (const row of rows) {
        copies += row.quantity;
        if (row.isFoil) foilCopies += row.quantity;
        if (!isPricedField(row.tcgMarket) && !isPricedField(row.tcgLow)) {
            tcgUnpricedCopies += row.quantity;
        }
        const trend = foilAware(row, 'cmTrendFoil', 'cmTrend');
        const low = foilAware(row, 'cmLowFoil', 'cmLow');
        if (!isPricedField(trend) && !isPricedField(low)) {
            cmUnpricedCopies += row.quantity;
        }
    }

    const ranked = [];
    for (const row of rows) {
        const value = sourceValue(row, source, headline);
        if (!isPricedField(value)) continue;
        ranked.push({
            printingId: row.printingId,
            name: row.name,
            finish: row.finish,
            quantity: row.quantity,
            contribution: Number(value) * row.quantity,
        });
    }
    ranked.sort((a, b) => {
        if (b.contribution !== a.contribution) return b.contribution - a.contribution;
        const byName = a.name.localeCompare(b.name);
        if (byName !== 0) return byName;
        return a.printingId.localeCompare(b.printingId);
    });

    return {
        copies,
        distinctPrintings: rows.length,
        foilCopies,
        regularCopies: copies - foilCopies,
        tcgMarket: sumField(rows, (r) => r.tcgMarket),
        tcgLow: sumField(rows, (r) => r.tcgLow),
        cmTrend: sumField(rows, (r) => foilAware(r, 'cmTrendFoil', 'cmTrend')),
        cmLow: sumField(rows, (r) => foilAware(r, 'cmLowFoil', 'cmLow')),
        tcgUnpricedCopies,
        cmUnpricedCopies,
        topPrintings: ranked.slice(0, 5),
    };
}

File: apps/mobile/lib/core/logic/binder_value_snapshot.dart
import '../models/app_settings.dart';
import '../models/binder_entry.dart';

/// Observed catalog field vs missing. Numeric `0` is unpriced, never a summand.
bool isPricedField(double? value) =>
    value != null && value != 0 && !value.isNaN;

double? _firstPriced(Iterable<double?> values) {
  for (final value in values) {
    if (isPricedField(value)) return value;
  }
  return null;
}

/// How [BinderValueSnapshot.topPrintings] pick a per-copy value.
enum BinderValueHeadline {
  /// Mobile chip: [PriceSource] fallbacks (`Pricing.value` with 0 treated as
  /// unpriced).
  pricingValue,

  /// Web Binder header: TCGplayer Market only, no Low/Mid/High fallback.
  tcgMarketOnly,
}

class MoneyTotal {
  const MoneyTotal({
    required this.amount,
    required this.pricedCopies,
    required this.unpricedCopies,
  });

  /// `null` when [pricedCopies] is 0 — display as unpriced, never `$0.00`.
  final double? amount;
  final int pricedCopies;
  final int unpricedCopies;
}

class BinderValueTopPrinting {
  const BinderValueTopPrinting({
    required this.printingId,
    required this.name,
    required this.finish,
    required this.quantity,
    required this.contribution,
  });

  final String printingId;
  final String name;
  final String finish;
  final int quantity;
  final double contribution;
}

/// Quantity-weighted Binder totals for the value overlay. Input must be Binder
/// rows only (not Want List).
class BinderValueSnapshot {
  const BinderValueSnapshot({
    required this.copies,
    required this.distinctPrintings,
    required this.foilCopies,
    required this.regularCopies,
    required this.tcgMarket,
    required this.tcgLow,
    required this.cmTrend,
    required this.cmLow,
    required this.tcgUnpricedCopies,
    required this.cmUnpricedCopies,
    required this.topPrintings,
  });

  final int copies;
  final int distinctPrintings;
  final int foilCopies;
  final int regularCopies;
  final MoneyTotal tcgMarket;
  final MoneyTotal tcgLow;
  final MoneyTotal cmTrend;
  final MoneyTotal cmLow;
  final int tcgUnpricedCopies;
  final int cmUnpricedCopies;
  final List<BinderValueTopPrinting> topPrintings;
}

/// One Printing's observed fields for the snapshot. Contract tests build this
/// from JSON; the Binder screen maps [BinderEntry] through [fromBinderEntry].
class BinderValueRow {
  const BinderValueRow({
    required this.printingId,
    required this.name,
    required this.finish,
    required this.isFoil,
    required this.quantity,
    this.tcgMarket,
    this.tcgLow,
    this.tcgMid,
    this.tcgHigh,
    this.cmTrend,
    this.cmLow,
    this.cmAvg,
    this.cmTrendFoil,
    this.cmLowFoil,
    this.cmAvgFoil,
  });

  final String printingId;
  final String name;
  final String finish;
  final bool isFoil;
  final int quantity;
  final double? tcgMarket;
  final double? tcgLow;
  final double? tcgMid;
  final double? tcgHigh;
  final double? cmTrend;
  final double? cmLow;
  final double? cmAvg;
  final double? cmTrendFoil;
  final double? cmLowFoil;
  final double? cmAvgFoil;

  factory BinderValueRow.fromBinderEntry(BinderEntry entry) {
    final card = entry.card;
    return BinderValueRow(
      printingId: card.id,
      name: card.name,
      finish: card.subTypeName ?? (card.isFoil ? 'Foil' : 'Normal'),
      isFoil: card.isFoil,
      quantity: entry.quantity,
      tcgMarket: card.tcgMarket,
      tcgLow: card.tcgLow,
      tcgMid: card.tcgMid,
      tcgHigh: card.tcgHigh,
      cmTrend: card.cmTrend,
      cmLow: card.cmLow,
      cmAvg: card.cmAvg,
      cmTrendFoil: card.cmTrendFoil,
      cmLowFoil: card.cmLowFoil,
      cmAvgFoil: card.cmAvgFoil,
    );
  }

  factory BinderValueRow.fromJson(Map<String, dynamic> json) {
    double? n(String key) {
      final v = json[key];
      if (v == null) return null;
      return (v as num).toDouble();
    }

    return BinderValueRow(
      printingId: json['printingId'] as String,
      name: json['name'] as String,
      finish: json['finish'] as String,
      isFoil: json['isFoil'] as bool? ?? false,
      quantity: (json['quantity'] as num).toInt(),
      tcgMarket: n('tcgMarket'),
      tcgLow: n('tcgLow'),
      tcgMid: n('tcgMid'),
      tcgHigh: n('tcgHigh'),
      cmTrend: n('cmTrend'),
      cmLow: n('cmLow'),
      cmAvg: n('cmAvg'),
      cmTrendFoil: n('cmTrendFoil'),
      cmLowFoil: n('cmLowFoil'),
      cmAvgFoil: n('cmAvgFoil'),
    );
  }

  double? get foilAwareTrend =>
      isFoil ? _firstPriced([cmTrendFoil, cmTrend]) : _firstPriced([cmTrend]);

  double? get foilAwareLow =>
      isFoil ? _firstPriced([cmLowFoil, cmLow]) : _firstPriced([cmLow]);

  double? sourceValue(PriceSource source, BinderValueHeadline headline) {
    if (headline == BinderValueHeadline.tcgMarketOnly) {
      return _firstPriced([tcgMarket]);
    }
    switch (source) {
      case PriceSource.tcgplayer:
        return _firstPriced([tcgMarket, tcgLow, tcgMid, tcgHigh]);
      case PriceSource.cardmarket:
        if (isFoil) {
          return _firstPriced([cmTrendFoil, cmTrend, cmAvg, cmLow]);
        }
        return _firstPriced([cmTrend, cmAvg, cmLow]);
    }
  }
}

MoneyTotal _sumField(List<BinderValueRow> rows, double? Function(BinderValueRow) field) {
  var amount = 0.0;
  var pricedCopies = 0;
  var unpricedCopies = 0;
  var anyPriced = false;
  for (final row in rows) {
    final value = field(row);
    if (isPricedField(value)) {
      amount += value! * row.quantity;
      pricedCopies += row.quantity;
      anyPriced = true;
    } else {
      unpricedCopies += row.quantity;
    }
  }
  return MoneyTotal(
    amount: anyPriced ? amount : null,
    pricedCopies: pricedCopies,
    unpricedCopies: unpricedCopies,
  );
}

/// Builds overlay totals from Binder rows. Drops quantity &lt; 1.
BinderValueSnapshot buildBinderValueSnapshot(
  Iterable<BinderValueRow> entries, {
  PriceSource source = PriceSource.tcgplayer,
  BinderValueHeadline headline = BinderValueHeadline.pricingValue,
}) {
  final rows = [
    for (final row in entries)
      if (row.quantity >= 1) row,
  ];

  var copies = 0;
  var foilCopies = 0;
  var tcgUnpricedCopies = 0;
  var cmUnpricedCopies = 0;
  for (final row in rows) {
    copies += row.quantity;
    if (row.isFoil) foilCopies += row.quantity;
    final tcgPriced =
        isPricedField(row.tcgMarket) || isPricedField(row.tcgLow);
    if (!tcgPriced) tcgUnpricedCopies += row.quantity;
    if (!isPricedField(row.foilAwareTrend) && !isPricedField(row.foilAwareLow)) {
      cmUnpricedCopies += row.quantity;
    }
  }

  final ranked = <BinderValueTopPrinting>[];
  for (final row in rows) {
    final value = row.sourceValue(source, headline);
    if (!isPricedField(value)) continue;
    ranked.add(BinderValueTopPrinting(
      printingId: row.printingId,
      name: row.name,
      finish: row.finish,
      quantity: row.quantity,
      contribution: value! * row.quantity,
    ));
  }
  ranked.sort((a, b) {
    final byContrib = b.contribution.compareTo(a.contribution);
    if (byContrib != 0) return byContrib;
    final byName = a.name.compareTo(b.name);
    if (byName != 0) return byName;
    return a.printingId.compareTo(b.printingId);
  });

  return BinderValueSnapshot(
    copies: copies,
    distinctPrintings: rows.length,
    foilCopies: foilCopies,
    regularCopies: copies - foilCopies,
    tcgMarket: _sumField(rows, (r) => r.tcgMarket),
    tcgLow: _sumField(rows, (r) => r.tcgLow),
    cmTrend: _sumField(rows, (r) => r.foilAwareTrend),
    cmLow: _sumField(rows, (r) => r.foilAwareLow),
    tcgUnpricedCopies: tcgUnpricedCopies,
    cmUnpricedCopies: cmUnpricedCopies,
    topPrintings: ranked.take(5).toList(growable: false),
  );
}

BinderValueSnapshot snapshotForBinder(
  Iterable<BinderEntry> entries, {
  required PriceSource source,
  BinderValueHeadline headline = BinderValueHeadline.pricingValue,
}) {
  return buildBinderValueSnapshot(
    [
      for (final entry in entries)
        if (!entry.isWanted) BinderValueRow.fromBinderEntry(entry),
    ],
    source: source,
    headline: headline,
  );
}
